VultisigApp/VultisigApp/Services/CryptoPriceService.swift

```diff
@@ -1,120 +1,124 @@
-        do {
-            struct Response: Codable {
-                struct Data: Codable {
-                    struct Attributes: Codable {
-                        let image_url: String?
-                        let price_usd: String?
-                    }
-                    let attributes: Attributes
-                }
-                let data: [Data]
-            }
-            
-            let response: Response = try await Utils.fetchObject(from: Endpoint.fetchTokensInfo(
-                network: chain.coingeckoId,
-                addresses: [contractAddress])
-            )
-            
-            if let response = response.data.first {
-                if let priceRate = response.attributes.price_usd {
-                    return (image_url: response.attributes.image_url, coingecko_coin_id: nil, price_usd: Double(priceRate))
-                }
+
+    func resolveSources(coins: [Coin]) -> ResolvedSources {
+        var providerIds: [String] = []
+        var contracts: [Chain: [String]] = [:]
+
+        for coin in coins {
+            switch RateProvider.cryptoId(for: coin) {
+            case .priceProvider(let id):
+                providerIds.append(id)
+            case .contract(let id):
+                contracts[coin.chain, default: []].append(id)
             }
-        } catch {
-            print(error.localizedDescription)
-            return (image_url: nil, coingecko_coin_id: nil, price_usd: nil)
         }
-        
-        return (image_url: nil, coingecko_coin_id: nil, price_usd: nil)
+
+        return ResolvedSources(providerIds: providerIds, contracts: contracts)
     }
-    
-    func getTokenPrice(coin: Coin) async -> Double {
-        
-        // Those tokens are the ones in the vault, so we should cache them if not cached
-        let vaultTokens = await getAllTokenPricesCoinGecko()
-        let vaultPrice = vaultTokens?.prices[coin.contractAddress]?[SettingsCurrency.current.rawValue.lowercased()]
-        
-        guard let price = vaultPrice else {
-            let prices = await fetchCoingeckoTokenPrices(contractAddresses: [coin.contractAddress], chain: coin.chain)
-            return prices?.prices[coin.contractAddress]?[SettingsCurrency.current.rawValue.lowercased()] ?? .zero
+
+    func fetchPrices(ids: [String]) async throws {
+        let idsQuery = ids
+            .filter { !$0.isEmpty }
+            .joined(separator: ",")
+
+        let currencies = SettingsCurrency.allCases
+            .map { $0.rawValue }
+            .joined(separator: ",")
+
+        let url = Endpoint.fetchCryptoPrices(
+            ids: idsQuery,
+            currencies: currencies
+        )
+
+        let (data, _) = try await URLSession.shared.data(from: url)
+        let response = try JSONDecoder().decode([String: [String: Double]].self, from: data)
+
+        try await RateProvider.shared.save(rates: mapRates(response: response))
+    }
+
+    func fetchPrices(contracts: [String], chain: Chain) async throws {
```

Comment:
Previously , we cache the price for a while , with this new implementation , rates are stored in database , and how often do we refresh the rates?

Reply:
We're refreshing rates in same places where we refreshing balances. Check BalanceService
If we fetch balances for whole value - fetching all rates
For single coin - fetching rate for this coin only